rbx_types/src/terrain.rs

```diff
@@ -0,0 +1,548 @@
+use std::{
+    collections::{BTreeMap, HashMap},
+    convert::TryFrom,
+    fmt,
+};
+
+use crate::Vector3;
+
+/// Size of a chunk. Chunks are cubes, so this is the length/width/height.
+const CHUNK_SIZE: i32 = 2i32.pow(5);
+
+/// Coordinates of a chunk or a voxel. For internal use.
+// Can't use Vector3int16; we need a 32 bit integer.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Default, PartialOrd, Ord, Hash)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+pub struct TerrainVec {
+    pub x: i32,
+    pub y: i32,
+    pub z: i32,
+}
+
+impl TerrainVec {
+    pub fn from_vec3(i: Vector3) -> Self {
+        Self {
+            x: i.x as i32,
+            y: i.y as i32,
+            z: i.z as i32,
+        }
+    }
+
+    pub fn new(x: i32, y: i32, z: i32) -> Self {
+        Self { x, y, z }
+    }
+}
+
+/// Coordinates of a `Voxel` inside of a `Chunk`, which is a grid of 4 units in world space.
+/// This is inside of a grid of 32^3 voxels per chunk.
+#[repr(transparent)]
+#[derive(Debug, PartialEq, Eq, Default, Clone, Copy, Hash)]
+#[cfg_attr(
+    feature = "serde",
+    derive(serde::Serialize, serde::Deserialize),
+    serde(transparent)
+)]
+pub struct VoxelCoordinates(TerrainVec);
+
+impl VoxelCoordinates {
+    /// Constructs a new `VoxelCoordinates` object.
+    #[inline]
+    pub fn new(x: i32, y: i32, z: i32) -> Self {
+        const VOXEL_MAX: i32 = CHUNK_SIZE - 1;
+        let (x, y, z) = (
+            x.clamp(0, VOXEL_MAX),
+            y.clamp(0, VOXEL_MAX),
+            z.clamp(0, VOXEL_MAX),
+        );
+        Self(TerrainVec::new(x, y, z))
+    }
+
+    /// Constructs a new `VoxelCoordinates` object from a Vector3.
+    #[inline]
+    pub fn from_vec3(i: Vector3) -> Self {
+        Self(TerrainVec::from_vec3(i))
+    }
+}
+
+/// Coordinates of a `Chunk` in chunk space, which is a grid of 128 units in world space.
+/// Relevant for usage with a `Terrain` object. Inside a grid of 524,288^3 chunks per world.
+#[repr(transparent)]
+#[derive(Debug, PartialEq, Eq, Default, Clone, Copy, PartialOrd, Ord, Hash)]
+#[cfg_attr(
+    feature = "serde",
+    derive(serde::Serialize, serde::Deserialize),
+    serde(transparent)
+)]
+pub struct ChunkCoordinates(TerrainVec);
+
+impl ChunkCoordinates {
+    /// Constructs a new `ChunkCoordinates` object.
+    #[inline]
+    pub fn new(x: i32, y: i32, z: i32) -> Self {
+        // Terrain isn't accepted by the engine with voxels past 2^23.
+        const CHUNK_MAX: i32 = 2i32.pow(23) / CHUNK_SIZE;
+        let (x, y, z) = (
+            x.clamp(-CHUNK_MAX, CHUNK_MAX),
+            y.clamp(-CHUNK_MAX, CHUNK_MAX),
+            z.clamp(-CHUNK_MAX, CHUNK_MAX),
+        );
+        Self(TerrainVec::new(x, y, z))
+    }
```

Comment:
```suggestion
    pub fn new(x: i32, y: i32, z: i32) -> Self {
        // Terrain isn't accepted by the engine with voxels past 2^23.
        const CHUNK_MAX: i32 = 2i32.pow(23) / CHUNK_SIZE;

        Self(TerrainVec::new(
            x.clamp(-CHUNK_MAX, CHUNK_MAX),
            y.clamp(-CHUNK_MAX, CHUNK_MAX),
            z.clamp(-CHUNK_MAX, CHUNK_MAX),
        ))
    }
```
Same as the other `new` function.